### PR TITLE
Redux updates to handle availability

### PR DESCRIPTION
**Pull request by firejake308**
"Squashed" all of the Redux updates for availability input into one separate PR. Now that I think about it, I think one of the benefits of Redux is supposed to be isolation of logic from UI, which I guess is coming in handy now.

"Squash" is in quotes because I actually just used `git checkout frontend/availability redux/`, which has effectively the same result.

### CODE REVIEW

File: autoscheduler/frontend/src/types/Availability.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+export default interface Availability {
+  dayOfWeek: number;
+  startTimeHours: number;
+  startTimeMinutes: number;
+  endTimeHours: number;
+  endTimeMinutes: number;
+  available: AvailabilityType;
+// eslint-disable-next-line semi
```
Comment (ryan-conn) [inline autoscheduler/frontend/src/types/Availability.ts]: Put this in `eslintrc` rules instead:
```
  "semi": "off",
  "@typescript-eslint/semi": ["error"],
```
I got the rules from [this](https://github.com/typescript-eslint/typescript-eslint/issues/123) page

File: autoscheduler/frontend/src/types/Availability.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+export enum AvailabilityType {
+  NONE, BUSY
+}
+
+export default interface Availability {
+  dayOfWeek: number;
```
[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/types/Availability.ts]: Good point. Sometimes I get confused as to whether weeks start at Monday or Sunday, so I'll probably add something in the `types` folder for that

File: autoscheduler/frontend/src/redux/reducers.ts [hunk 3]
```diff
@@ -61,10 +65,120 @@ function courseCards(
...
+              }
+              return avsList.filter((av) => av !== oldAv).concat(newAv);
+            }
+            // if they're different types, then set the new one to the old one's borders
+            // TODO
+          }
+        }
+        // if no overlap
+        return avsList.concat(oldAv);
+      }, []).concat(overlapsFound ? [] : initNewAv);
```
[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/redux/reducers.ts]: Unfortunately, that would be bad practice with Redux, since `.push()` would mutate the argument array. I'm not sure why mutations are so bad, but the Redux guide recommended avoiding them, which is why I have the `concat`s

File: autoscheduler/frontend/src/tests/Redux.test.ts [hunk 4]
```diff
@@ -183,3 +188,152 @@ test('Updates course card boolean field', () => {
...
+    available: AvailabilityType.BUSY,
+    dayOfWeek: 2,
+    startTimeHours: 7,
+    startTimeMinutes: 0,
+    endTimeHours: 13,
+    endTimeMinutes: 42,
+  }]);
+});
+
+test('Deletes availbility', () => {
```
[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/tests/Redux.test.ts]: I'm telling you, availability is just too long of a word to type correctly

File: autoscheduler/frontend/src/types/Availability.ts (hunk 1 above)
Comment (firejake308) [reply autoscheduler/frontend/src/types/Availability.ts]: Oh, okay, so replace the normal rule with the TS one. Thanks!

File: autoscheduler/frontend/src/redux/reducers.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryan-conn) [reply autoscheduler/frontend/src/redux/reducers.ts]: If I'm not mistaken, redux only wants you to avoid mutating any part of the state (so that time travel is possible) so since `avsList` is mutating the accumulator of the reduce function rather than part of the state, it should be find to use `push()`, since calling this with the same state/action will always result in the same value

[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/redux/reducers.ts]: そうです。ちがったので、ごめんなさい。

[the remark this answers is not in the document]
Comment (ryan-conn) [reply autoscheduler/frontend/src/redux/reducers.ts]: 謝っても、許せない。アデルさんはこのプロジェクトの不名誉だ。

冗談だけｗｗ

File: autoscheduler/frontend/src/tests/Redux.test.ts [hunk 5]
```diff
@@ -183,3 +188,152 @@ test('Updates course card boolean field', () => {
...
+    dayOfWeek: 2,
+    time1: 8 * 60 + 0,
+    time2: 12 * 60 + 0,
+  }));
+  const intermediateState = store.getState().availability;
+  store.dispatch(deleteAvailability({
+    available: AvailabilityType.BUSY,
+    dayOfWeek: 2,
+    time1: 8 * 60 + 0,
+    time2: 12 * 60 + 0,
```
Comment (gannonprudhomme) [inline autoscheduler/frontend/src/tests/Redux.test.ts]: Since you're using the same data twice, I'd say create the availability dict in arrange. Also think this will help with clarity of the test in general, since you'll see that you're adding and deleting the same object w/o having to check all of the individual fields.

Comment (gannonprudhomme) [reply autoscheduler/frontend/src/tests/Redux.test.ts]: Honestly I think this same sentiment applies to a bunch of the tests. For tests where the dicts aren't the same, extract out the constants that you're setting in every availability object (i.e. `dayOfWeek` and `available`).

Also, do it for times in the cases that they're small differences of each other (since the specific times in most cases don't really matter, it's their relation to the other times that do, i.e. for testing merging).

File: autoscheduler/frontend/src/redux/actions.ts [hunk 6]
```diff
@@ -114,3 +118,64 @@ export function updateCourseCard(index: number, courseCard: CourseCardOptions):
...
+  return {
+    type: DELETE_AVAILABILITY,
+    availability,
+  };
+}
+
+/**
+ * Updates the availability matching the given availability args. Enter the unchanged
+ * time as time1 and the changed time as time2
+ * @param availability
```
Comment (gannonprudhomme) [inline autoscheduler/frontend/src/redux/actions.ts]: Could you add something like `Called for updating the bounds of availbility by dragging the edges` or something like that? Just like whenever it'll be called to give some context

File: autoscheduler/frontend/src/tests/Redux.test.ts [hunk 7]
```diff
@@ -183,3 +188,152 @@ test('Updates course card boolean field', () => {
...
+    available: AvailabilityType.BUSY,
+    dayOfWeek: 2,
+    startTimeHours: 8,
+    startTimeMinutes: 0,
+    endTimeHours: 13,
+    endTimeMinutes: 42,
+  }]);
+});
+
+test('Doesn\'t merge non-overlapping availabilities', () => {
```
[the remark this answers is not in the document]
Comment (gannonprudhomme) [reply autoscheduler/frontend/src/tests/Redux.test.ts]: This would also us to nest describes so we can add a bunch of tests for a specific function without having to include the function in the test name.

File: autoscheduler/frontend/src/redux/reducers.ts [hunk 8]
```diff
@@ -61,10 +65,120 @@ function courseCards(
...
+              };
+              if (overlapsFound === 2) {
+                return avsList.slice(0, oldOverlapIdx)
+                  .concat(avsList.slice(oldOverlapIdx + 1))
+                  .concat(newAv);
+              }
+              return avsList.filter((av) => av !== oldAv).concat(newAv);
+            }
+            // if they're different types, then set the new one to the old one's borders
+            // TODO
```
[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/redux/reducers.ts]: Hmm, that´s space for a possible future functionality that would allow the user to set times that they would *prefer* not to have classes. I'd like to leave the comment there to make it easier to figure out in case we ever get around to adding that.

File: autoscheduler/frontend/src/redux/reducers.ts [hunk 9]
```diff
@@ -61,10 +65,120 @@ function courseCards(
...
+        // if no overlap
+        return avsList.concat(oldAv);
+      }, []).concat(overlapsFound ? [] : initNewAv);
+    }
+    case DELETE_AVAILABILITY:
+      // filters the availability list for the availabiltiy matching the action args
+      return state.filter((av) => av.available !== action.availability.available
+        || av.dayOfWeek !== action.availability.dayOfWeek
+        || av.startTimeHours * 60 + av.startTimeMinutes !== action.availability.time1
+        || av.endTimeHours * 60 + av.endTimeMinutes !== action.availability.time2);
```
[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/redux/reducers.ts]: Yes, it took me a second, but there actually is a reason. In the case of `DELETE_AVAILABILITY`, it is known that the argument `time1` refers to the start time and `time2` refers to the end time. However, for `UPDATE_AVAILABILITY`, `time1` could match either the start time or the end time. 

Therefore, since this giant conditional will return `false` iff `avl` does not match the arguments, in the case of deleting, finding a single un-matching time means the whole availability does not match, but in the case of updating, it must be verified that neither start nor end time matches in order to safely move onto the next availability.

File: autoscheduler/frontend/src/redux/reducers.ts [hunk 10]
```diff
@@ -61,10 +65,120 @@ function courseCards(
...
+    case UPDATE_AVAILABILITY: {
+      const { time1, time2 } = action.availability;
+      return state.map((av) => ((av.available !== action.availability.available
+        || av.dayOfWeek !== action.availability.dayOfWeek
+        || (av.startTimeHours * 60 + av.startTimeMinutes !== time1
+        && av.endTimeHours * 60 + av.endTimeMinutes !== time1))
```
[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/redux/reducers.ts]: Actually, no, the `UPDATE_AVAILABILITY` action uses `time1` as the known, *un*changed time to find which availability it is that you want to update, and `time2` is the other time, which *has* been changed. Therefore, `time2` is irrelevant when you're looking for the right availabiltiy to update, which is what this coditional is doing.

This information is documented in `actions.ts` above `updateAction()`, but if you'd like, I could add some comments here as well

File: autoscheduler/frontend/src/redux/reducers.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (gannonprudhomme) [reply autoscheduler/frontend/src/redux/reducers.ts]: Yeah I figured as much. In that case the comment on 126 is fine, but I'd say remove the `TODO`, as I generally like to use it as a in-progress indicator (plus just don't like committing them in general, `pylint` throws errors for them on the backend too)

**Comment by firejake308** [comment]
@gannonprudhomme I agree that organizing Redux files by the slice of the Redux state handled makes more sense, but I think it would be best done in a subsequent PR, just to prevent further bloat to this one.

File: autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    store.dispatch(addAvailability(availability1));
+    store.dispatch(addAvailability(availability2));
+
+    // assert
+    expect(store.getState().availability).toEqual(
+      [argsToAvailability(availability1), argsToAvailability(availability2)],
+    );
+  });
+
+  test('are deleted', () => {
```
Comment (gannonprudhomme) [inline autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: While I can't think of any other scenarios for testing delete, I think it'd be better if this was also in its own describe and the actual test described the exact scenario you're testing, as reading it as a glance I'm not really sure what situation you're testing. How you set up the tests for merge availability is perfect in this regard.

This also applies to update, where I think it would especially benefit from explaining the scenario you're testing in the test name. 

Basically I'm trying to find the middle ground of how Jest tests are setup versus my normal naming scheme (coming from Java/C#) of `SystemUnderTest_ScenarioWe'reTesting_ExpectedResult`

File: autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+      };
+
+      // act
+      store.dispatch(addAvailability(availability1));
+      store.dispatch(addAvailability(availability2));
+
+      // assert
+      expect(store.getState().availability).toEqual([{
+        available: AvailabilityType.BUSY,
+        dayOfWeek: DayOfWeek.WED,
```
[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: I had originally left those in there to wxplicitly show the final state, but I guess it's irrelevant, so I've gone ahead and replaced all of them with `dummyArgs`

File: autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+      store.dispatch(addAvailability(availability3));
+
+      // assert
+      expect(store.getState().availability).toEqual([{
+        available: AvailabilityType.BUSY,
+        dayOfWeek: DayOfWeek.WED,
+        startTimeHours: 7,
+        startTimeMinutes: 0,
+        endTimeHours: 13,
+        endTimeMinutes: 42,
```
[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: This is a trivial change for me to make, but for improved understanding of good test formatting, could you explain why you arrange the tests this way? Is the `assert` step supposed to be pretty minimal?

[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: Also, would you prefer to have them called `expected` or something more descriptive like `mergedAvailability`?

File: autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx (hunk 11 above)
Comment (firejake308) [reply autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: How about the top-level describe blocks names the system under test (pretty much the same as the file name), the middle describe identifies the expected result, and the test is named for the exact scenario tested?

Comment (gannonprudhomme) [reply autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: Yeah I think that's perfect

File: autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (gannonprudhomme) [reply autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: Yeah so I generally just like to have a consistent structure with the tests I write, so in the case of a test failing we can quickly open the test and see what the expected value is should be, rather than looking at the `assert` and seeing the variable we're comparing to (i.e. to manually find what the expected is). Basically, it's to save time, and in the case of naming it `mergedAvailability` I think it's probably just as descriptive as naming it `expected`, given that we're testing availability merging.

This is a really trivial / small thing and honestly I don't think it matters *that* much, and at the end of the day it's just personal preference. I just think having all of our tests that are doing some computation and comparing the result to the expected value name the expected value be `expected` gives a good consistency to our tests.

[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: Aight done

**Comment by firejake308** [comment]
> Creating 2 blocks and then one that overlaps both will merge twice
This problem has been resolved. The others are UI issues that will be resolved in the next PR. If y'all agree with the changes made in this last commit, then this should be ready to go

**Comment by firejake308** [comment]
Arrrghhh wait nvm there's still a bug

**Comment by firejake308** [comment]
Ok this time it's done for real for real

File: autoscheduler/frontend/src/redux/reducers.ts [hunk 14]
```diff
@@ -61,10 +102,117 @@ function courseCards(
...
+    case ADD_AVAILABILITY:
+    case MERGE_AVAILABILITY: {
+      // check for overlaps between the availability to be added and pre-existing ones
+      let overlapsFound = 0; // counts merging overlaps only
+      let overlapIdx = -1;
+
+      const initNewAv = action.type === ADD_AVAILABILITY
+        ? argsToAvailability(action.availability)
+        : state[state.length - 1];
+      let newAv = initNewAv;
```
Comment (gannonprudhomme) [inline autoscheduler/frontend/src/redux/reducers.ts]: Shouldn't we set `newAv` to what `initNewAv` is? Since this is the only time we use `initNewAv`

File: autoscheduler/frontend/src/redux/reducers.ts [hunk 15]
```diff
@@ -61,10 +102,117 @@ function courseCards(
...
+                startTimeHours: avWithEarlierStart.startTimeHours,
+                startTimeMinutes: avWithEarlierStart.startTimeMinutes,
+                endTimeHours: newEndHrs,
+                endTimeMinutes: newEndMins,
+              };
+              if (overlapsFound === 2) {
+                avsList.splice(oldOverlapIdx, 1);
+                return avsList;
+              }
+              return avsList.filter((av) => av !== oldAv);
```
Comment (gannonprudhomme) [inline autoscheduler/frontend/src/redux/reducers.ts]: I'd also say add comments to this block just explaining the difference between returning in the `if` statement versus the default `.filter` return

File: autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+        },
+      ];
+
+      // act
+      store.dispatch(addAvailability(availability1));
+      store.dispatch(addAvailability(availability2));
+      store.dispatch(addAvailability(availability3));
+      store.dispatch(addAvailability(availability4));
+      store.dispatch(addAvailability(availability5));
+      store.dispatch(updateAvailability(availability5New));
```
Comment (gannonprudhomme) [inline autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: What's the point of updating this availability? Ideally we wouldn't use other functions in our unit tests, as if `updateAvailability` is failing, this test will fail as well. Is this to get around not being able to mock mouse inputs?

**Comment by gannonprudhomme** [comment]
Also, how did you remove the review requests? I didn't know that was an option lol

File: autoscheduler/frontend/src/redux/reducers.ts (hunk 14 above)
Comment (firejake308) [reply autoscheduler/frontend/src/redux/reducers.ts]: Fixed

File: autoscheduler/frontend/src/redux/reducers.ts (hunk 15 above)
Comment (firejake308) [reply autoscheduler/frontend/src/redux/reducers.ts]: There is no difference. They have both been replaced with `return avsList;`

File: autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx (hunk 16 above)
Comment (firejake308) [reply autoscheduler/frontend/src/tests/ReduxAvailability.test.tsx]: No, but it replicates how the UI would do it. When the user first clicks, it adds a 0-minute-long availability, then updates it as the user drags, and then merges when the user lets go.

File: autoscheduler/frontend/src/redux/reducers.ts [hunk 17]
```diff
@@ -149,9 +149,8 @@ function availability(
                 endTimeHours: newEndHrs,
                 endTimeMinutes: newEndMins,
               };
-              if (overlapsFound >= 2) {
+              if (overlapsFound === 2) {
```
[the remark this answers is not in the document]
Comment (firejake308) [reply autoscheduler/frontend/src/redux/reducers.ts]: This was actually a very good comment to make. Originally, when an overlap occurred, the merged availability would be inserted at the index of the pre-existing availability that the new one overlapped with. However, this would insert the availability multiple times if the new availability overlapped with multiple pre-existing availabilities, leading to the bug that Ryan reported. The splicing was how I got rid of the duplicates. Originally, I had assumed that it would not be possible for more than 2 overlaps to occur, so I didn't allow splicing for that, but after Ryan's bug report, I realized I might have to splice more often than anticipated. That is roughly why I changed it from `=== 2` to `>= 2`. However, this led to too much splicing, accidentally deleting unrelated availabilities. To fix this, I removed splicing altogether, and instead, I just insert the new, merged availability at the very end. Doing so eliminates the need for splicing altogether, and also simplifies the code somewhat.
